### PR TITLE
CADC-10809 - add HDF5 support.

### PR DESCRIPTION
**Pull request by SharonGoliath**

### CODE REVIEW

File: caom2utils/caom2utils/caom2blueprint.py [hunk 1]
```diff
@@ -3167,70 +3592,85 @@ def _has_data_array(self, header):
...
 
-        :param keywords: Keywords to add to a CAOM2 set.
-        :param current: Existing CAOM2 entity with a keywords attribute.
-        :param to_set: A CAOM2 entity with a keywords attribute.
-        """
-        if keywords:
-            if isinstance(keywords, set):
-                to_set.keywords.update(keywords)
+class HDF5Parser(ContentParser):
+
```
Comment (andamian) [inline caom2utils/caom2utils/caom2blueprint.py:3597]: Probably it's premature, but a description similar to the one for the `FitsParser` would be nice.

Comment (andamian) [inline caom2utils/caom2utils/caom2blueprint.py:3597]: Probably it's premature, but a description similar to the one for the `FitsParser` would be nice.

File: caom2utils/caom2utils/caom2blueprint.py [hunk 2]
```diff
@@ -3167,70 +3592,85 @@ def _has_data_array(self, header):
...
+    def __init__(
+        self, obs_blueprint, uri, local_f_name, find_roots_here='/sitedata'
+    ):
+        import h5py
```
Comment (andamian) [inline caom2utils/caom2utils/caom2blueprint.py:3601]: Please add note as to why this import is located here.

Comment (andamian) [inline caom2utils/caom2utils/caom2blueprint.py:3601]: Please add note as to why this import is located here.

File: caom2utils/caom2utils/tests/data/taos_h5file/20220201T200117/taosii.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    return ra
+
+
+def get_target_position_cval2(base):
+    ra_ignore, dec = _get_target_position(base)
+    return dec
+
+
+def _get_target_position(base):
+    import logging
```
Comment (andamian) [inline caom2utils/caom2utils/tests/data/taos_h5file/20220201T200117/taosii.py:26]: These imports can be moved up at the start of the file I think

Comment (andamian) [inline caom2utils/caom2utils/tests/data/taos_h5file/20220201T200117/taosii.py:26]: These imports can be moved up at the start of the file I think

File: doc/user/script_description.md [hunk 4]
```diff
@@ -2,13 +2,16 @@
...
 
 `augment` works by creating or augmenting a CAOM2 Observation record, which can then be loaded via the CADC service.
 
-`augment` creates the Observation record using information contained in a FITS file. The python module `fits2caom2`, from the python package `caom2utils`,
-examines the FITS file and uses a blueprint, embodied in an instance of the ObsBlueprint class, to define default values, override values, and mappings to augment the FITS header. The keywords and values in the augmented FITS header are then used to fill in corresponding CAOM2 entities and attributes.
+`augment` creates the Observation record using information contained in a FITS file. The python module `caom2blueprint`, from the python package `caom2utils`,
+examines the FITS or HDF5 file and uses a blueprint, embodied in an instance of the ObsBlueprint class, to define default values, override values, and mappings to augment the file 
+metadata content. The augmented keywords and values are then 
+checked for consistency with `astropy`'s [WCS package](https://docs.astr.
```
Comment (andamian) [inline doc/user/script_description.md:13]: Missing end of line

Comment (andamian) [inline doc/user/script_description.md:13]: Missing end of line

**Review by andamian — COMMENTED** [review]
This was a quick review

File: caom2utils/caom2utils/caom2blueprint.py [hunk 5]
```diff
@@ -1442,12 +1445,19 @@ def is_fits(value):
...
+        return ObsBlueprint.needs_lookup(value) and value[0] == 'BINTABLE'
 
     @staticmethod
     def is_function(value):
-        return (not ObsBlueprint.is_fits(value) and isinstance(value, str)
-                and isinstance(value, str) and '(' in value and ')' in value)
+        return (not ObsBlueprint.needs_lookup(value) and isinstance(value, str)
+                # and isinstance(value, str) and '()' in value)
+                and isinstance(value, str) and '(' in value and ')' in value
+                and '/' not in value)
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:1461]: `doc` added to function.

File: caom2utils/caom2utils/caom2blueprint.py [hunk 6]
```diff
@@ -1584,19 +1949,16 @@ def augment_plane(self, plane, artifact_uri):
...
             plane.artifacts[artifact_uri] = artifact
-        self.augment_artifact(artifact)
+        self.augment_artifact(artifact, 0)
         self.logger.debug(
-            'End generic CAOM2 plane augmentation for {}.'.format(
-                artifact_uri))
+            f'End CAOM2 plane augmentation for {artifact_uri}.')
 
-    def augment_artifact(self, artifact):
+    def augment_artifact(self, artifact, index):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:1962]: `:param` description added. It's used in the specializing classes.

File: caom2utils/caom2utils/caom2blueprint.py (hunk 1 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:3610]: Made a start on the `doc` string.

File: caom2utils/caom2utils/caom2blueprint.py (hunk 2 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:3614]: Added.

File: caom2utils/caom2utils/tests/data/taos_h5file/20220201T200117/taosii.py (hunk 3 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/tests/data/taos_h5file/20220201T200117/taosii.py:26]: They're vestigial, and shouldn't be there at all. I' still in the "learning HDF5" phase, and trying to figure out errors.

File: doc/user/script_description.md (hunk 4 above)
Comment (SharonGoliath) [reply doc/user/script_description.md:13]: Reworded a bit. Hopefully it's more correct now.

File: caom2utils/caom2utils/caom2blueprint.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:1471]: Added `doc`.

File: caom2utils/caom2utils/caom2blueprint.py [hunk 7]
```diff
@@ -2125,6 +2158,13 @@ def _execute_external_instance(self, value, key, extension):
             logging.debug(
                 'Key {} calculated value of {} using {}'.format(
                     key, result, value))
+        except ValueError as e2:
+            # DB 23-03-22
+            # Anything that you can do to make the CAOM2 record creation fail
+            # in this case of bad WCS metadata would be useful. Use
+            # ValueError because that happens to be what astropy is throwing
+            # for a SkyCoord construction failure.
+            raise Caom2Exception(e2)
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/caom2blueprint.py:2167]: Added `:raise` parameter.

**Review by andamian — APPROVED** [review]

**Pull request merged**